Add daily cron job to disable inactive FreeIPA users

jsf9k opened the pull request.

jsf9k committed:
Add cron job that runs daily and disables inactive users

jsf9k committed:
Add comment indicating possible future improvement

jsf9k committed:
Extract substring in a more verbose manner

The templating foo in cloud_init.tf does not seem to like the
expression g=${g:${#tmp} + 1}.

jsf9k committed:
Remove unnecessary line

dav3r commented on cloud-init/write_disable_inactive_users_cron_job.tpl.yml:
```diff
@@ -0,0 +1,88 @@
+      ###
+      # Determine the date 45 days ago, in a format that ldapsearch likes
+      ###
+      distant_past=$(date --date="$(date) -45 days" +%Y%m%d%H%M%SZ)
```
It would be nice to make the number of days (45, in our case) into a terraform variable for increased flexibility.

jsf9k replied:
See d0cbd30.

dav3r replied:
👍

dav3r commented on cloud-init/write_disable_inactive_users_cron_job.tpl.yml:
```diff
@@ -0,0 +1,88 @@
+      for user in $users_to_disable
+      do
+          # The ipa user-disable command may fail if the user is already
+          # disabled, but that's not really en error condition for us.
```
Can you think of any other ways the `ipa user-disable` might fail that would be unintentionally masked by `set +o errexit`?  Is this something worth worrying about?

jsf9k replied:
I don't think it's something worth worrying about.  I can't think of other reasons right now (although I'm sure they exist) but if the `ipa user-disable` fails the reason why will be apparent in the cron logs.

All `set -o errexit` does is exit the bash script if any command returns an exit code indicating a failure.  We're temporarily turning that behavior off so that the loop can continue on to disabling the next inactive user.

I'll add a comment clarifying this.

jsf9k replied:
See 174ffa3.

dav3r replied:
Thanks- makes sense!

dav3r reviewed (approved):
Looks good, though I left a few comments for your consideration.

I also investigated to see if I could find any better ways to deploy this shell script so that it could be checked by our pre-commit hooks, but I came up empty. 😢

felddy reviewed (commented):
It'd be nice if this wasn't in cloudinit so we could get some shell-check love.

The script would need to get the `domain` variable at runtime.  This would also make it tolerant to domain changes. (which will probably never happen)

I couldn't find a k-command to spit out the realms, but something like this maybe?:
```console
[mark.feldhousen@ipa ~]$ grep "default_realm" /etc/krb5.conf | awk '{print $3}'
COOL.CYBER.DHS.GOV
```

With that change it could be installed back in ansible land.

jsf9k committed:
Add a variable for the number of days before an account is labeled inactive

Thanks to @dav3r for the suggestion.

jsf9k committed:
Add a comment clarifying why it's OK to disable errexit

This is referring to the spot in the disable_user script where we
actually disable users.

jsf9k commented:
> It'd be nice if this wasn't in cloudinit so we could get some shell-check love.
> 
> The script would need to get the `domain` variable at runtime. This would also make it tolerant to domain changes. (which will probably never happen)
> 
> I couldn't find a k-command to spit out the realms, but something like this maybe?:
> 
> ```
> [mark.feldhousen@ipa ~]$ grep "default_realm" /etc/krb5.conf | awk '{print $3}'
> COOL.CYBER.DHS.GOV
> ```
> 
> With that change it could be installed back in ansible land.

I am out of time now, but I will look into determining the domain at runtime when I have a chance.

felddy reviewed (approved):
Opened an issue for future consideration.  Approved.

jsf9k commented:
Closing, since the solution in cisagov/ansible-role-freeipa-server#5 is preferred.
